### DISCUSSION TITLE
Toggle Midi Binding should act on the inital state of the preset control.

### DISCUSSION THREAD

**Discussion by rerdavies** [rerdavies/pipedal#158] [General]
But.....
I seem to miss the meaning of 'toggle'.
I thought that each time a same midi cc event (only CC nr and data value no meaning) occurred the status would change.
SO starting an preset in which a AMP is on and pressing the footswitch would result in OFF.
But I noticed that sending a CC with value 127 only switches it on and never off.

So what is the story?
And is it possible to make such a 'toggle'? So a CC nr, data value unimportant would switch the switch to the opposite state...
This is important for me since after changing the preset I do not want to remember whether I have to step on the foot pedal switch once or twice to get the opposite state...

Cheers,

Maarten

_Originally posted by @maarthome in https://github.com/rerdavies/pipedal/issues/152#issuecomment-2278276220_

**Comment by rerdavies** [comment]
The Toggle action is triggered when the control switches from OFF to ON state. The purpose of this behaviour is to avoid toggling on every CC message if the MIDI controller is some kind of continous controller. 

What you actually need to do is send a CC nr, 127, followed by a cc nr,00, so that the next time you send the messages, PiPedal sees an OFF to ON transition.

I have opened FEATURE REQUEST to add a (?) button that will provide a help dialog with an detailed description of the two options. (high priority, will be added before  v1.2 leaves beta).

It's a confusing feature. Hopefully a help dialog will go some way to address that.

**Reply by maarthome** [reply]
Hi Robin,

To avoid long discussions... is it possible to create such functionality that ONE CC that is sent, 'toggles' the switch status to the opposite state?
In order to do this I assume the current status must be known. A XOR function? 10 -> on and 11 -> off...
So the momentatry MIDI pulse (the CC) is always 1 independent from the actual CC data... CC34 with data 127 is treated the same as CC34 with any other value (0:126)

Cheers

**Comment by rerdavies** [comment]
That's what toggle does. It toggles the current value of the LV2 control,
regardless of the value(s) of the MIDI cc message.

The feature is confusing enough, without introducing a third option. The
existing functionality works well with Midi footswitch (where the
footswitch has been configured as momentary on the  footswitch controller).
Note that the Midi footswtich controller has no idea what the state of the
LV2 control is, so this is the best solution.

Is this for your GPIO project? Just send two midi messages (127,0), when
the switch goes on. (Or send a CCC when the switch goes off).

**Comment by rerdavies** [comment]
Premature send. :-/

Anything is possible. And the toggle function already does an XOR of the
LV2 state.  But, I am reluctant to create a third option at this point,
given how confusing it already is. Can you explain what problem you're
trying to solve?

It could be that I'm mistaken about what MIDI pad controls and foot
controller switches do when they're configured to send CCs.

**Reply by rerdavies** [reply]
Dawning comprehension.... Let me look at what my MIDI controllers do for CC messages. I may be entirely wrong about this. Hold that thought.

**Reply by rerdavies** [reply]
No. I'm not entirely wrong. 

So I pulled up the configuration editor for my Alesis controller.   When a pad is configured as a CC controller, I'm offered the following options:

- The CC Number
- The CC value to send when pressed. I set it to 127.
- The CC value to send when not pressed. I set it to 0.

There's no option to send just one message.

Please describe what controller you're using, and why you need  a toggle on one message only.

**Comment by rerdavies** [comment]
Could you please send me the output of the following commands:

    ls -l /var/pipedal
   ls -l /var/pipedal/audio_uploads
   ls -l /var/pipedal/audio_uploads/CabIR

**Comment by maarthome** [comment]
He Robin,

Thanks for looking into this.
I am pretty sure that I tested a lot of settings... :-)
My simple cheap controller can send only one cc message when pressed... And that really does not work....
But I did not try the CC 127 and CC 00 after one momentary switch activation..... but tomorrow morning :-)

PS I do not grasp how the setting of the ON/OFF status of the switch is not used by the code...??? I do not think it can work correctly...??

Cheers, I am off to bed...
Maarten

**Comment by maarthome** [comment]
We have two discussions in one eh.... conversation.... 🤣🤣

**Comment by rerdavies** [comment]
We do. Sorry. I'm going nuts trying to find the tail ends of our separate
conversations. lol

This is why I *should *be more insistent on routing discussions on
unrelated issues to another discussion or bug report. But Inever do. :-(

In the meantime, two issues:

1.  Please  send me the dump of permissions on the folders I asked for.
2. Please describe what your controller is, and what configuration options
for the midi device you have available to you.

**Comment by rerdavies** [comment]
What controller are you using? I'll pull the manuals and see what it's
capable of.

**Comment by maarthome** [comment]
Good day 👍👍👍
MIDI switching  WORKS!! Wow... I am happy!!
I guess with all the chaotic experiments I missed sending a CC 127 and immediately a CC 0...... (probabely because I miss the logic of this...)
But it works. Set it to Toggle and define sending two messages for 1 CC, the first with data 127 and the second with data 0.

I am going for lunh with my youngest daughter in Utrecht so I have to wait until late.... and I am so eager to play..... he, but what is the nicest of the two ... 🤣

Thanks Robin for being so patient (and polite 👍😊)

By the way, here is the directory listing you asked for:

-rw-rw-r-- 1 pipedal_d pipedal_d   198 Aug 11 08:50 AudioConfig.json
drwxrwsr-x 7 pipedal_d pipedal_d  4096 Aug  9 14:46 audio_uploads
drwxrwsr-x 2 pipedal_d pipedal_d  4096 Aug  8 21:25 home
-rw-rw-r-- 1 pipedal_d pipedal_d   251 Aug 11 08:53 JackChannelSelection.json
drwxrwsr-x 2 pipedal_d pipedal_d 20480 Aug  9 10:19 plugin_presets
drwxrwsr-x 2 pipedal_d pipedal_d  4096 Aug 11 08:58 presets
-rw-rw-r-- 1 pipedal_d pipedal_d   485 Aug 10 01:06 SystemMidiBindings.json

Have a great day and enjoy!!
Maarten

**Reply by rerdavies** [reply]
Just to summarize:

When set to toggle, pipedal waits for a MIDI CC with an ON value. When it triggers, it toggles the value of the current plugin control (regardless of what the value of the MIDI CC was).. PiPedal will not trigger the toggle again, until it has seen a MIDI CC with an OFF value. So you either need to configure your MIDI controller to send an ON message when your MIDI controller's switch or pad is pressed, and an OFF message when it's released, or send both an ON message immediately followed by  an OFF message when your MIDI controller's pad or switch is pressed. 

I think an ON messages is a CC with a value greater than or equal to 64, and an OFF message is a CC with a value less than 64, but I'd have to check the sources to confirm that. That's what it should be. If it isn't let me know, and a fix the code so that it is.

If your MIDI controller doesn't support either of these options, it may be easier just to configure your controller to send MIDI NOTES instead of MIDI CCs. Invariably, MIDI controllers will send a NOTE ON when the switch or pad is pressed, and a NOTE OFF when the pad is released. Which is what PiPedal expects.

**Reply by maarthome** [reply]
Hi Robin,

Think there is a misunderstanding now.
It works now. I do exactly what you suggest: I send two CC messages with the same CC number. Only the first time data value is 127 and the second time it is 0. To me this it not logic, but who cares when you know how  to make it work 👍.
(Logic to me would be sending only one CC with a random value)

Thanks agian for clarifying it for me.... 
(perhaps it should be in the help text.... or manual..)

Cheers,
Maarten

..

**Comment by rerdavies** [comment]
That looks correct.

1. Could you also send me output for the following two commands.

    ls -l /var/pipedal/audio_uploads

and

    ls -l /var/pipedal/audio_uploads/CabIR

2. Reconfigure your audio buffer settings:

The default buffer settings for PiPedal are 64 byte buffers, 2 buffers.
This seems to have been a bad choice.

in PiPedal settings, configure your audio buffer as 64 bytes, 3 buffers!

After doing so, keep an eye on the audio status. If you are experiencing
excessive XRuns (more than one or two on initial configuration), or if the
audio status goes to Stopped, let me know.

The message you received earlier....  something about not being able to
SetPatchParam .... should  instead read (and will be revised to)"Unable to
perform the current operation because audio processing has stopped".

**Comment by maarthome** [comment]
Hi Robin,

Here are the directory listings. (one is quite long 😁)
  ls -l /var/pipedal/audio_uploads:

drwxrwsr-x 3 pipedal_d pipedal_d  4096 Aug  9 11:12 AIDA-X.lv2
drwxrwsr-x 2 pipedal_d pipedal_d  4096 Aug 10 22:44 CabIR
lrwxrwxrwx 1 root      pipedal_d    45 Aug 11 10:22 MOD-ConvolutionLoader.lv2 -> /var/pipedal/audio_uploads/ReverbImpulseFiles
drwxrwsr-x 2 pipedal_d pipedal_d 32768 Aug  8 21:53 NeuralAmpModels
drwxrwsr-x 2 pipedal_d pipedal_d 12288 Aug 11 10:19 ReverbImpulseFiles
lrwxrwxrwx 1 root      pipedal_d    18 Aug 11 22:46 zeroconvo.lv2 -> ReverbImpulseFiles

and...
 ls -l /var/pipedal/audio_uploads/CabIR



-rw-rw-r-- 1 pipedal_d pipedal_d 28880 Aug  9 14:35  412-MRSHLL-1960AHW-G12H_MIX-01.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28880 Aug  9 14:35  412-MRSHLL-1960AHW-G12H_MIX-02.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28880 Aug  9 14:35  412-MRSHLL-1960AHW-G12H_MIX-03.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28880 Aug  9 14:35  412-MRSHLL-1960AHW-G12H_MIX-04.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28880 Aug  9 14:35  412-MRSHLL-1960AHW-G12H_MIX-05.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28880 Aug  9 14:35  412-MRSHLL-1960AHW-G12H_MIX-06.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28880 Aug  9 14:35  412-MRSHLL-1960AHW-G12H_MIX-07.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28880 Aug  9 14:35  412-MRSHLL-1960AHW-G12H_MIX-08.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28880 Aug  9 14:35  412-MRSHLL-1960AHW-G12H_MIX-09.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28880 Aug  9 14:35  412-MRSHLL-1960AHW-G12H_MIX-10.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28880 Aug  9 14:35  412-MRSHLL-1960AHW-G12H_MIX-11.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28880 Aug  9 14:35  412-MRSHLL-1960AHW-G12H_MIX-12.wav
lrwxrwxrwx 1 pipedal_d pipedal_d    58 Aug 10 22:44 '80s UK 001.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/80s UK 001.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    58 Aug 10 22:44 '80s UK 002.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/80s UK 002.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    58 Aug 10 22:44 '80s US 001.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/80s US 001.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    58 Aug 10 22:44 '80s US 002.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/80s US 002.wav'
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_CG_C12_0in_CapEdge.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_CG_C12_0in_Cap.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_CG_C12_0in_ConeEdge.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_CG_C12_0in_Cone.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_CG_C12_2in_CapEdge.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_CG_C12_2in_Cap.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_CG_C12_2in_ConeEdge.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_CG_C12_2in_Cone.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_CG_C12_6in_CapEdge.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_CG_C12_6in_Cap.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_CG_C12_6in_ConeEdge.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_CG_C12_6in_Cone.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_NV_C12_0in_CapEdge.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_NV_C12_0in_Cap.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_NV_C12_0in_ConeEdge.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_NV_C12_0in_Cone.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_NV_C12_2in_CapEdge.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_NV_C12_2in_Cap.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_NV_C12_2in_ConeEdge.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_NV_C12_2in_Cone.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_NV_C12_6in_CapEdge.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_NV_C12_6in_Cap.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_NV_C12_6in_ConeEdge.wav
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  Bogie412CB65_NV_C12_6in_Cone.wav
lrwxrwxrwx 1 pipedal_d pipedal_d    61 Aug 10 22:44 'Modern UK 001.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Modern UK 001.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    61 Aug 10 22:44 'Modern UK 002.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Modern UK 002.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    61 Aug 10 22:44 'Modern US 001.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Modern US 001.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    61 Aug 10 22:44 'Modern US 002.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Modern US 002.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    61 Aug 10 22:44 'Modern US 003.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Modern US 003.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    61 Aug 10 22:44 'Modern US 004.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Modern US 004.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    57 Aug 10 22:44 'Radio 001.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Radio 001.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    59 Aug 10 22:44 'Speaker 001.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Speaker 001.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    59 Aug 10 22:44 'Speaker 002.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Speaker 002.wav'
-rw-rw-r-- 1 pipedal_d pipedal_d 28848 Aug 10 22:43  testie.wav
lrwxrwxrwx 1 pipedal_d pipedal_d    62 Aug 10 22:44 'Vintage UK 001.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Vintage UK 001.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    62 Aug 10 22:44 'Vintage UK 002.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Vintage UK 002.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    62 Aug 10 22:44 'Vintage US 001.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Vintage US 001.wav'
lrwxrwxrwx 1 pipedal_d pipedal_d    62 Aug 10 22:44 'Vintage US 002.wav' -> '/usr/lib/lv2/ToobAmp.lv2/impulseFiles/CabIR/Vintage US 002.wav'
root@pipedal:/var/pipedal/audio_uploads#

**Comment by maarthome** [comment]
Ow... forgot, enjoy your evening/night....
Cheers,

Maarten

**Comment by rerdavies** [comment]
Ok. That looks correct, except for the extra file

    lrwxrwxrwx 1 root pipedal_d 18 Aug 11 22:46 zeroconvo.lv2 ->
ReverbImpulseFiles

which is harmless. It's good to see that the factory IRs have shown up.
That definitely indicates that the new re-permissioning code has improved
things, and that you probably won't see the Updload failed/Can't create
file problem anymore.

You initially reported a couple of symptoms. Could you please describe to
me what the symptoms are on the current version.

If you see the dialog message about not being able to set a parameter, that
actually means that your audio thread has crashed. If you look at the audio
status at the top of the settings page (or the bottom left corner of the
screen, if you haven't turned it off), it will say "Stopped". That's my
fault. I changed the buffer defaults to 64x2 on the last release.
Unfortunately this works very badly, causing so many overruns that
eventually the audio thread does crash. Changing your audio buffers to 64x3
will fix this problem. I *think *32x3 will be fine too, although you might
get sporadic XRuns at that setting with the NAM plugin.

Waiting for your response.

- R
